### PR TITLE
Fix to #23761 - UseRelationalNulls causes subquery to include NOT IN (NULL, x)

### LINKED ISSUE #23761: UseRelationalNulls causes subquery to include NOT IN (NULL, x)
This query produces an invalid WHERE when UseRelationalNulls = true for nullable int.
Incorrect SQL producted: **AND [a].[StatusCode] NOT IN (NULL, 3)**
```
await db.ProjectModels
  .Where(p => p.ProjectId == projectid)
  .Select(p => new ScoreRemainingActivitiesInput
  {
      RemainingActivities = p.Activities
          .Where(a => a.TaskType != TaskTypeEnum.TT_WBS && a.TaskType != TaskTypeEnum.TT_LOE)
          .Where(a => a.StatusCode != null && a.StatusCode != StatusCodeEnum.TK_Complete)
          .Count()
  }).FirstOrDefaultAsync();
```

SQL when UseRelationalNulls = false.
```
SELECT TOP (1) (
    SELECT COUNT(*)
    FROM [ActivityModels] AS [a]
    WHERE (
        ([p].[ProjectId] = [a].[ProjectId])
        AND (
          [a].[TaskType] NOT IN (6, 3)
          OR [a].[TaskType] IS NULL
          )
        )
      AND (
        ([a].[StatusCode] <> 3)
        AND [a].[StatusCode] IS NOT NULL
        )
    ) AS [RemainingActivities]
FROM [ProjectModels] AS [p]
WHERE [p].[ProjectId] = '1A66B859-F096-4595-8B30-D127D8CA6E0E'
```

SQL when UseRelationalNulls = true.
```
SELECT TOP (1) (
    SELECT COUNT(*)
    FROM [ActivityModels] AS [a]
    WHERE (
        ([p].[ProjectId] = [a].[ProjectId])
        AND [a].[TaskType] NOT IN (6, 3)
        )
      AND [a].[StatusCode] NOT IN (NULL, 3)
    ) AS [RemainingActivities]
FROM [ProjectModels] AS [p]
WHERE [p].[ProjectId] = '1A66B859-F096-4595-8B30-D127D8CA6E0E'
```

EF Core version: 5.0.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 5.0
Operating system: Windows 10
IDE: Visual Studio 2019 16.8.3

### PR DESCRIPTION
**Pull request by maumar**
We do optimization which combines comparisons based on the same property into IN and NOT IN. Problem is that if the comparisons contain nulls, we add those into the list of IN values, generating queries like  a IN (1, 2, NULL). In normal circumstances, during null semantics processing we extract these nulls and produce correct IS NULL / IS NOT NULL calls, however when useRelationalNulls is enabled, we don't run null semantics visitor and therefore don't "fix" the IN expressions.

Fix is to only apply the initial optimization for c# null semantics.

Fixes #23761

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs [hunk 1]
```diff
@@ -14,6 +15,8 @@ namespace Microsoft.EntityFrameworkCore.Query
     /// <inheritdoc />
     public class RelationalQueryTranslationPostprocessor : QueryTranslationPostprocessor
     {
+        private readonly bool _useRelationalNulls = false;
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs:18]:
```suggestion
        private readonly bool _useRelationalNulls;
```

File: test/EFCore.Relational.Specification.Tests/Query/NullSemanticsQueryTestBase.cs [hunk 2]
```diff
@@ -1530,13 +1530,99 @@ public virtual async Task Nested_CompareTo_optimized(bool async)
...
+
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual async Task Multiple_equality_comparisons_including_null_comparison_work_for_relational_null_semantics(bool async)
+        {
+            var ctx = CreateContext(useRelationalNulls: true);
+
+            var expected = ctx.Entities1.AsEnumerable().Where(e => e.NullableIntA == 1 || e.NullableIntA != null).ToList();
+            ClearLog();
+            var query = ctx.Entities1.Where(e => e.NullableIntA == 1 || e.NullableIntA == null);
```
Comment (smitpatel) [inline test/EFCore.Relational.Specification.Tests/Query/NullSemanticsQueryTestBase.cs:1562]: expected and query are different?

File: src/EFCore.Relational/Query/Internal/SqlExpressionSimplifyingExpressionVisitor.cs [hunk 3]
```diff
@@ -287,10 +298,17 @@ private Expression SimplifySqlBinary(SqlBinaryExpression sqlBinaryExpression)
...
                         }
 
-                        return _sqlExpressionFactory.In(
-                            leftCandidateInfo.ColumnExpression,
-                            _sqlExpressionFactory.Constant(resultArray, leftCandidateInfo.TypeMapping),
-                            leftCandidateInfo.OperationType == ExpressionType.NotEqual);
+                        // we can't do IN optimizations when using relational null semantics
+                        // a != 1 && a != null would be converted to a NOT IN (1, null), which never returns any results
+                        // we need to keep it in the original form so that a != null gets converted to a IS NOT NULL instead
+                        if (!_useRelationalNulls)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/SqlExpressionSimplifyingExpressionVisitor.cs:304]: comparison + comparison is also fine to collapse as long as the comparison is not to null.

File: src/EFCore.Relational/Query/Internal/SqlExpressionSimplifyingExpressionVisitor.cs [hunk 4]
```diff
@@ -272,6 +274,15 @@ private Expression SimplifySqlBinary(SqlBinaryExpression sqlBinaryExpression)
                             leftValue = leftCandidateInfo.ConstantValue;
                             rightValue = rightCandidateInfo.ConstantValue;
                             resultArray = UnionCollections((IEnumerable)leftValue, (IEnumerable)rightValue);
+
+                            if (_useRelationalNulls)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/SqlExpressionSimplifyingExpressionVisitor.cs:278]: This block should not be needed. Rather the the generation later should take care of excluding case where it is not possible to generate (perhaps an early return is also fine.)

Comment (smitpatel) [reply src/EFCore.Relational/Query/Internal/SqlExpressionSimplifyingExpressionVisitor.cs:278]: Like if identify that there is a comparison (so other 2 blocks than this), which compares to null, we return from there only as it is not collapsible. For now we can ignore different level of nesting and ordering of terms. e.g. `(a == 1 || a == null) || a == 2`

File: test/EFCore.Relational.Specification.Tests/Query/NullSemanticsQueryTestBase.cs [hunk 5]
```diff
@@ -1530,13 +1530,99 @@ public virtual async Task Nested_CompareTo_optimized(bool async)
...
+            Assert.Equal(expected.Count, result.Count);
+        }
+
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual async Task Multiple_contains_calls_get_combined_into_one_for_relational_null_semantics(bool async)
+        {
+            var ctx = CreateContext(useRelationalNulls: true);
+
+            var expected = ctx.Entities1.AsEnumerable().Where(e => new int?[] { 1, 2, 3 }.Contains(e.NullableIntA)).ToList();
```
Comment (smitpatel) [inline test/EFCore.Relational.Specification.Tests/Query/NullSemanticsQueryTestBase.cs:1574]: add some tests which also use relational nulls but no comparison to null and still optimize like C# nulls.

**Review by smitpatel — APPROVED** [review]
